### PR TITLE
feat(cast): adding config support to cast

### PR DESCRIPTION
**Pull request by SusheendharVijay**
## Motivation
Let cast commands consume items from config file.  Addressing #689

### CODE REVIEW

**Comment by SusheendharVijay** [comment]
Hey @mattsse, I've made quite a few changes so I thought I'll get your feedback

changes: 
- added config support for all commands that needed just `rpc_url`
- added config support for `cast find-block`, `cast estimate`
- changed the `etherscan_api_key` field of `EthereumOpts` to `etherscan_key` to avoid collision with Config. 
- added `impl_eth_data_provider` macro to implement `Provider` trait for commands that use `EthereumOpts`. This also makes it easy to implement commands in the future. 
- made macros more verbose to skip unnecessary imports.

**Comment by mds1** [comment]
How exactly does adding config support to cast work? To be honest I'm not sure I fully see the use case for this, so some specific questions I have:

1. What is the full set of config params that can apply to cast? ETH_RPC_URL and ETHERSCAN_API_KEY, but anything else?
2. Will this automatically use settings in a `foundry.toml` file if I run `cast` from a directory containing a config file? I'm not sure how I feel about that being the default, and it would probably be a breaking change for a lot people who rely on behavior similar to seth, so if that's the plan we should consider disabling it by default and letting users enable it
3. Does this also respect a global `foundry.toml` I define in my home directory, and is that always fallen back to if a local `foundry.toml` isn't found?

**Comment by dmfxyz** [comment]
Outside of @mds1 questions, I have some (perhaps paranoid) hesitations about supporting an API secret in an oft committed configuration file.

**Comment by SusheendharVijay** [comment]
Adding config support is just merging the cli commands to the `Config` object, just like `forge` does. Everything upto generating the `Config` file is unchanged. It's the same way forge consumes `Config` items.  

Config object is created by merging options set in (ascending order of precedence):
-  global foundry.toml 
- local foundry.toml
- env variables
- cli inputs. (this part is what I'm adding for `cast`)

This is how it's in `seth` too right?   This makes `cast` have the same behaviour as `seth`. From docs: 
image


1. Apart from `eth_rpc_url`  and `etherscan_api_key` there's `sender` (from), `chain_id` and `block_number`. (not sure about how useful chain_id and block_number are)
2. Yes, it will use settings from local file over global file, `seth` has the same behaviour afaict. If no option is set locally, then the global option is used. The only difference is that local `foundry.toml` is created by default, but with `seth` we have to create a `.sethrc`, but none of these options are set by default so pretty much the same. 
3.  Yes it respects global `foundry.toml` and it will fall back to it in case there's no local config file.
4. @dmfxyz yeah this is a problem, way too many private keys on github lmao. This issue exists with `seth` too. While writing this pr I wanted it to be just like `seth` so there's no problem for people relying on `seth`'s behavior like @mds1 righlty mentioned.

Hopefully this makes sense, correct me If I'm wrong, thanks!

**Comment by mds1** [comment]
Thanks for the detailed answer @SusheendharVijay! I do like the use case you linked in #1160, e.g. `cast tx <hash> -p <profileName>`. Would this PR enable me to setup profiles with different RPC URLs and Etherscan API keys in `~/foundry.toml` and use them like that? 

However I still think we should be careful about changing the default cast behavior as part of this PR since doing so is a breaking change and people may have scripts that rely on the current behavior. We may want to put this behind a flag such as `--config`/`-c` until we do a larger cast refactor where breaking changes are ok, cc @onbjerg

**Comment by onbjerg** [comment]
> However I still think we should be careful about changing the default cast behavior as part of this PR since doing so is a breaking change and people may have scripts that rely on the current behavior. We may want to put this behind a flag such as `--config`/`-c` until we do a larger cast refactor where breaking changes are ok, cc @onbjerg

What part is a breaking change?

**Comment by mds1** [comment]
Sorry, to be more clear, it _may_ be a breaking change depending on the specific setup users have. The main thing I had in mind here is using the ETH_RPC_URL to fetch data in scripts, which now starts pulling from a different chain because of a local or global config file you may have. 

One example is a repo doing some cross-chain development where they have a `foundry.toml` pointing to an L2 RPC URL for tests, and a deploy script that uses `cast` with the ETH_RPC_URL env var to read some L1 data as inputs for their L2 deploy script. Their deploy script will now start reading L2 data when it previously was reading L1 data.

This is a made up example and it's possible there's zero repos for which this will be a breaking change, but wanted to call it out since some people rely heavily on seth for scripting and may have swapped it out for cast.

**Comment by SusheendharVijay** [comment]
Hey @mds1, I don't think it will break it here, since cast did not have config support, people had to use the ETH_RPC_URL env var, so since we are adding config support, which has lower precedence, it will be overidden by the env var. It seems like it will be backwards compatible. If a env var is set, it will be overriden for forge commands too, so there wouldn't be such scenarios. Hopefully I'm not dragging the discussion too long, I think it's better to be safe, just can't think of a scenario where it would break. You can foundryup and maybe test it out.

**Comment by sambacha** [comment]
> Hey @SusheendharVijay, here's what I'm seeing. 
> 
> 
> 
> I have `ETH_RPC_URL=myMainnetRpcUrl` in my `~/.zshrc`, and in my local project I added a `cast.sh` script that runs the following:
> 
> 
> 
> ```sh
> 
> LOCAL_RPC_URL=myMainnetRpcUrl
> 
> cast chain-id
> 
> cast chain-id --rpc-url $ETH_RPC_URL
> 
> cast chain-id --rpc-url $LOCAL_RPC_URL
> 
> ```
> 
> 
> 
> Then I added a `foundry.toml` that looks like this:
> 
> 
> 
> ```toml
> 
> [default]
> 
> eth-rpc-url="myOptimismRpcUrl"
> 
> ```
> 
> 
> 
> Then I did these tests:
> 
> 1. Using the latest version of foundry (not this PR). I ran `bash cast.sh` and all three commands returned a chain ID of 1, as expected.
> 
> 2. Then I ran `foundryup -P 1245` and confirmed the returned version was commit 00070eb from this PR. I ran `bash cast.sh` and all three commands returned a chain ID of 1. I'm not sure if this is expected, or because my `ETH_RPC_URL` env var overrode it
> 
> 3. I updated `cast.sh` to have `ETH_RPC_URL=` to unset that env var, and comments out `cast chain-id --rpc-url $ETH_RPC_URL` in our script because we know it will faill. Re-running `bash cast.sh` then shows some errors (I added some `echo` statements to more clearly show which commands the errors are from). From the below, it seems `cast chain-id` now fails and is not picking up the config var in the toml file
> 
> 
> 
> ```
> 
> bash cast.sh
> 
> 
> 
> About to run: cast chain-id
> 
> Error: 
> 
>    0: relative URL without a base
> 
> 
> 
> Location:
> 
>    cli/src/cast.rs:205
> 
> 
> 
> Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
> 
> Run with RUST_BACKTRACE=full to include source snippets.
> 
> 
> 
> About to run: cast chain-id --rpc-url $LOCAL_RPC_URL
> 
> 1
> 
> ```

Why are you running bash cash when you have zsh, jw?

**Comment by mds1** [comment]
> Why are you running bash cast when you have zsh, jw?

Mainly just out of habit to ensure consistency when other people run scripts since I assume bash is more common. Though I think zsh is actually now the default on macOS so maybe outdated reasoning / just habit at this point 🤷

**Comment by mds1** [comment]
Ah I see. I `export` the env var in zshrc so it's shared between shells—if it wasn't then `cast chain-id --rpc-url $ETH_RPC_URL` would have failed in the first test

**Comment by SusheendharVijay** [comment]
Hey @mds1, the `relative URL without a base` error happens when the url is wrong, so something's being consumed but the url isn't right for some reason, sure that the optimism url is correct? I'll re-run this same test locally, thanks!

**Comment by mds1** [comment]
I am sure the Optimism RPC URL is correct: if I copy it directly out of my `foundry.toml` and run `cast chain-id --rpc-url <copiedFromToml>`, it prints `10` as expected. Specifically I have:

```toml
[default]
eth-rpc-url="https://opt-mainnet.g.alchemy.com/v2/alchemyApiKeyHere"
```

Let me know if there's anything else you need from me to help debug!

**Comment by SusheendharVijay** [comment]
So my local tests seem to be working as expected, sharing details: 

foundry.toml (local)
```toml
[default]
eth-rpc-url = "https://opt-mainnet.g.alchemy.com/v2/nicetryhaha"
```

cast.sh:
```sh
export local_rpc=https://eth-mainnet.alchemyapi.io/v2/nope
export configCast=/home/sushi/code/crypto/eth/foundry/target/debug/cast
echo "Should give optimism chain-id: 10"
$configCast chain-id
echo "Should give eth-mainnet chain-id: 1"
$configCast chain-id --rpc-url $local_rpc
```
Output:
image


Maybe something unexpected in your environment? Any env variable set before calling `cast chain-id`? Maybe try listing env vars and checking. Thanks, this is fun 😄

**Comment by SusheendharVijay** [comment]
Oh wait I think when you set `ETH_RPC_URL=`, I think it's getting set as ' ', hence the invalid url error?

**Comment by mds1** [comment]
Ah, you are correct! Using `unset ETH_RPC_URL` instead now gives me the same results as your comment 🎉

Given this I would say this precedence seems safe and unlikely to break something. 

One note is that it's common to define an `ETH_RPC_URL` in `~/.zshrc` (or equivalent for other shells) since it makes cast UX much smoother. So the UX might be a bit confusing if people define an `eth-rpc-url` in the config file and it doesn't override their `ETH_RPC_URL`, but I think just documenting that in the foundry book and mentioning they'll need to `unset` is ok. (I'm using `ETH_RPC_URL` as a proxy for all params supported by this PR)

Thanks for putting up with all my comments! 😅

**Comment by SusheendharVijay** [comment]
> So the UX might be a bit confusing if people define an `eth-rpc-url` in the config file and it doesn't override their `ETH_RPC_URL`, but I think just documenting that in the foundry book and mentioning they'll need to `unset` is ok,

Great point, this should be mentioned, plus encouraging them to set it in the global toml file instead to skip the unset. Using the toml file over env vars seems better imo, more explicit



> Thanks for putting up with all my comments!

On the contrary, really apreciate you taking the time to stress test the pr!

**Comment by mds1** [comment]
> plus encouraging them to set it in the global toml file instead to skip the unset

Ah, does the global toml file override the env var? I don't think I realized that. That seems reasonable, but yea documenting the order of precedence somewhere would definitely help

> Using the toml file over env vars seems better imo, more explicit

Hmm, I'm on the fence here. I want to agree, but until the config file better supports env vars you have to hardcode an RPC URL with an API key into a file that gets committed, which isn't great. Once env vars are better supported, then I'd probably want to use config files in projects, but env vars globally for generic usage, for the better UX + simpler to quickly switch networks

**Comment by sambacha** [comment]
> > plus encouraging them to set it in the global toml file instead to skip the unset
> 
> 
> 
> Ah, does the global toml file override the env var? I don't think I realized that. That seems reasonable, but yea documenting the order of precedence somewhere would definitely help
> 
> 
> 
> > Using the toml file over env vars seems better imo, more explicit
> 
> 
> 
> Hmm, I'm on the fence here. I want to agree, but until the config file better supports env vars you have to hardcode an RPC URL with an API key into a file that gets committed, which isn't great. Once env vars are better supported, then I'd probably want to use config files in projects, but env vars globally for generic usage, for the better UX + simpler to quickly switch networks

Config varies substantially across deploys, code does not.

Having them set in foundry.toml is an anti pattern. You can load env using a makefile or justfile see this example/ https://github.com/sambacha/foundry-scripts/blob/master/justfile


env vars are granular controls, each fully orthogonal to other env vars. Foundry.toml is of a type of config that does not vary between deploys, and so is best done in the code. 

Env vars always take precedence over config defined values. 

This is a substantial change.

**Comment by sambacha** [comment]
Just want to be clear: this is a bad change: telling users to commit credential access in a configuration file is most certainly bad practice, and it also does not scale.

**Comment by dmfxyz** [comment]
> > So the UX might be a bit confusing if people define an `eth-rpc-url` in the config file and it doesn't override their `ETH_RPC_URL`, but I think just documenting that in the foundry book and mentioning they'll need to `unset` is ok,
> 
> Great point, this should be mentioned, plus encouraging them to set it in the global toml file instead to skip the unset. Using the toml file over env vars seems better imo, more explicit
> 
> > Thanks for putting up with all my comments!
> 
> On the contrary, really apreciate you taking the time to stress test the pr!

Unless I'm misunderstanding, I have to disagree. Environment variables should always take precedence over config files, even global ones. This is a very established pattern in development. Without it there's a risk of introducing some really weird behavior like:

`ETH_RPC_URL=MyEthRpcUrl cast send --private key ...` 
where you silently are still using the ETH_RPC_URL defined in the global config file despite having set the environment.

**Comment by SusheendharVijay** [comment]
Env vars still take precedence over all the config files!  Please refer to this: 

> Config object is created by merging options set in (ascending order of precedence):
> 
> * global foundry.toml
> * local foundry.toml
> * env variables
> * cli inputs. (this part is what I'm adding for `cast`)

This isn't a substantial change at all, in fact nothing changes for people using the current flow. I agree that it's risky to set api keys to the local config file, I just meant setting in global config over .zsrhc. In fact the current behaviour is exatly how seth behaves! Is it cool @sambacha/ @dmfxyz ? 

> Ah, does the global toml file override the env var? I don't think I realized that. That seems reasonable, but yea documenting the order of precedence somewhere would definitely help

No no it doesn't, I meant setting senstive keys in the global file instead of in .zshrc, that's all.

**Comment by dmfxyz** [comment]
Got it. Misunderstood the prior exchange 👍

**Comment by mds1** [comment]
Ah I also misunderstood. Contrary to my prior comment I agree env var always taking precedence makes sense 🙂

**Review by gakonst — COMMENTED** [review]
could we add some tests that confirm the config values are read correctly? here's an example https://github.com/gakonst/foundry/blob/master/cli/tests/it/config.rs#L296-L301

**Comment by onbjerg** [comment]
Terribly sorry about the merge conflicts I caused, I moved around a lot of the structs and changed a lot of help text to make this clearer when running `--help` 😞 Let me know if you need assistance rebasing/resolving conflicts

**Comment by SusheendharVijay** [comment]
That's no problem @onbjerg, I saw your issue, cleaning up `--help` was def needed, I'll resolve the conflicts and get back to you if any issues.

File: cli/src/cast.rs [hunk 1]
```diff
@@ -156,80 +159,79 @@ async fn main() -> eyre::Result<()> {
...
         Subcommands::BlockNumber { rpc_url } => {
+            let rpc_url = consume_config_rpc_url(rpc_url);
             let provider = Provider::try_from(rpc_url)?;
             println!("{}", Cast::new(provider).block_number().await?);
         }
 
-        Subcommands::Call(config_call) => {
-            let config: Config = (&config_call).into();
+        Subcommands::Call { address, sig, args, block, eth } => {
+            let config: Config = (&eth).into();
```
Comment (mattsse) [inline cli/src/cast.rs:185]:
```suggestion
            let config = Config::from(&eth);
```

File: cli/src/utils.rs [hunk 2]
```diff
@@ -129,6 +129,15 @@ pub fn parse_u256(s: &str) -> eyre::Result<U256> {
     Ok(if s.starts_with("0x") { U256::from_str(s)? } else { U256::from_dec_str(s)? })
 }
 
+/// Return `rpc-url` cli argument if given, or consume `eth-rpc-url` from foundry.toml. Default to
+/// `localhost:8545`
+pub fn consume_config_rpc_url(rpc_url: Option<String>) -> String {
+    match rpc_url {
```
Comment (mattsse) [inline cli/src/utils.rs:135]: use `if let Some() else {}` for Option

**Review by mattsse — APPROVED** [review]
smol nits

lgtm, wdyt @onbjerg ?

**Comment by onbjerg** [comment]
Am I understanding correctly that this only applies to RPC URL and Etherscan API key?

**Comment by SusheendharVijay** [comment]
> Am I understanding correctly that this only applies to RPC URL and Etherscan API key?

Yeah pretty much, also sender but that's not too commonly used.

**Review by onbjerg — APPROVED** [review]
Cool, lgtm!

**Review by mattsse — APPROVED** [review]

**Comment by sambacha** [comment]
> Env vars still take precedence over all the config files! Please refer to this:
> 
> > Config object is created by merging options set in (ascending order of precedence):
> > 
> > * global foundry.toml
> > * local foundry.toml
> > * env variables
> > * cli inputs. (this part is what I'm adding for `cast`)
> 
> This isn't a substantial change at all, in fact nothing changes for people using the current flow. I agree that it's risky to set api keys to the local config file, I just meant setting in global config over .zsrhc. In fact the current behaviour is exatly how seth behaves! Is it cool @sambacha / @dmfxyz ?
> 
> > Ah, does the global toml file override the env var? I don't think I realized that. That seems reasonable, but yea documenting the order of precedence somewhere would definitely help
> 
> No no it doesn't, I meant setting senstive keys in the global file instead of in .zshrc, that's all.
💯🥇

**Pull request merged**